### PR TITLE
MsgPack end-to-end

### PR DESCRIPTION
**Pull request by moozzyk**
Plugging in MsgPack protocol
Adding negotiation message to hub protocol

The following comes next/separately:
- TS client
- Version in negotiate
- Updating the protocol spec

### CODE REVIEW

File: src/Microsoft.AspNetCore.SignalR.Client/HubConnection.cs [hunk 1]
```diff
@@ -77,6 +78,12 @@ public HubConnection(IConnection connection, IHubProtocol protocol, ILoggerFacto
         public async Task StartAsync()
         {
             await _connection.StartAsync();
+
+            using (var memoryStream = new MemoryStream())
+            {
+                NegotiationProtocol.TryWriteProtocolNegotiationMessage(new NegotiationMessage(_protocol.GetType().Name), memoryStream);
+                await _connection.SendAsync(memoryStream.ToArray(), CancellationToken.None);
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR.Client/HubConnection.cs]: Failures here will most likely need to be handled

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/JsonHubProtocol.cs [hunk 2]
```diff
@@ -44,6 +44,8 @@ public JsonHubProtocol(JsonSerializer payloadSerializer)
             _payloadSerializer = payloadSerializer;
         }
 
+        public string Name { get => "Json"; }
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/JsonHubProtocol.cs]: `Name => "json"`

File: src/Microsoft.AspNetCore.SignalR/Internal/DefaultHubProtocolResolver.cs [hunk 3]
```diff
@@ -9,10 +10,17 @@ namespace Microsoft.AspNetCore.SignalR.Internal
 {
     public class DefaultHubProtocolResolver : IHubProtocolResolver
     {
-        public IHubProtocol GetProtocol(ConnectionContext connection)
+        public IHubProtocol GetProtocol(string protocolName, ConnectionContext connection)
         {
-            // TODO: Allow customization of this serializer!
-            return new JsonHubProtocol(new JsonSerializer());
+            switch (protocolName)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR/Internal/DefaultHubProtocolResolver.cs]: `ToLowerInvariant()` this shouldn't be case sensitive

**Comment by moozzyk** [comment]
f8d91b54d4b677b619d716f5003f2ab98b614829

**Pull request closed without merging**